### PR TITLE
Refactor override configs

### PR DESCRIPTION
**Pull request by grouma**
More towards #1427

`findBuildConfigOverrides` has a dependency on `watcher`. Refactor it out of `options`.

### CODE REVIEW

File: build_runner/lib/src/generate/watch_impl.dart [hunk 1]
```diff
@@ -69,8 +70,9 @@ Future<ServeHandler> watch(
       reader: reader,
       writer: writer,
       onLog: onLog);
+  overrideBuildConfig ??
```
Comment (natebosch) [inline build_runner/lib/src/generate/watch_impl.dart]: did you intend to use `??=`?

File: build_runner/lib/src/generate/build.dart [hunk 2]
```diff
@@ -80,12 +81,13 @@ Future<BuildResult> build(List<BuilderApplication> builders,
...
       onLog: onLog);
   var options = await BuildOptions.create(environment,
-      configKey: configKey,
       deleteFilesByDefault: deleteFilesByDefault,
       failOnSevere: failOnSevere,
       packageGraph: packageGraph,
       logLevel: logLevel,
       skipBuildScriptCheck: skipBuildScriptCheck,
+      overrideBuildConfig:
+          await findBuildConfigOverrides(packageGraph, configKey),
```
Comment (natebosch) [inline build_runner/lib/src/generate/build.dart]: does this mean we need to expose `findBuildConfigOverrides` from `build_runner_core`? Could `BuildOptions.create` call it instead?

File: build_runner/lib/src/generate/watch_impl.dart (hunk 1 above)
Comment (grouma) [reply build_runner/lib/src/generate/watch_impl.dart]: Yes. Fixed.

File: build_runner/lib/src/generate/build.dart (hunk 2 above)
Comment (grouma) [reply build_runner/lib/src/generate/build.dart]: No. The intent behind this PR is to allow `findBuildConfigOverrides` to remain in `build_runner`.

Comment (natebosch) [reply build_runner/lib/src/generate/build.dart]: ah, this is the `package_name.build.yaml` handling? I guess Flutter won't need to support that.

Comment (jakemac53) [reply build_runner/lib/src/generate/build.dart]: This is actually, `build.<config>.yaml` handling. The main issue is actually that we are trying to do as little yaml parsing as possible so the main api in core can only deal with `BuildConfig` objects.

The `--config` flag and semantics of it are something build_runner specific imo, and doesn't belong in core.

**Pull request merged**